mod/state-transition/pkg/core/deposits_validation.go

```diff
@@ -0,0 +1,158 @@
+]) validateGenesisDeposits(
+	st BeaconStateT,
+	deposits []DepositT,
+) error {
+	switch {
+	case sp.cs.DepositEth1ChainID() == spec.BartioChainID:
+		// Bartio does not properly validate deposits index
+		// We skip checks for backward compatibility
+		return nil
+
+	case sp.cs.DepositEth1ChainID() == spec.BoonetEth1ChainID:
+		// Boonet inherited the bug from Bartio and it may have added some
+		// validators before we activate the fork. So we skip validation
+		// before fork activation
+		return nil
+
+	default:
+		if _, err := st.GetEth1DepositIndex(); err == nil {
+			// there should not be Eth1DepositIndex stored before
+			// genesis first deposit
+			return errors.Wrap(
+				ErrDepositMismatch,
+				"Eth1DepositIndex should be unset at genesis",
+			)
+		}
+		if len(deposits) == 0 {
+			// there should be at least a validator in genesis
+			return errors.Wrap(
+				ErrDepositsLengthMismatch,
+				"at least one validator should be in genesis",
+			)
+		}
+		for i, deposit := range deposits {
+			// deposit indices should be contiguous
+			if deposit.GetIndex() != math.U64(i) {
+				return errors.Wrapf(
+					ErrDepositIndexOutOfOrder,
+					"genesis deposit index: %d, expected index: %d",
+					deposit.GetIndex().Unwrap(), i,
+				)
+			}
+		}
+		return nil
+	}
+}
+
+func (sp *StateProcessor[
+	_, _, _, BeaconStateT, _, DepositT,
+	_, _, _, _, _, _, _, _, _, _, _,
+]) validateNonGenesisDeposits(
+	st BeaconStateT,
+	deposits []DepositT,
+) error {
+	slot, err := st.GetSlot()
+	if err != nil {
+		return fmt.Errorf(
+			"failed loading slot while processing deposits: %w", err,
+		)
+	}
+	switch {
+	case sp.cs.DepositEth1ChainID() == spec.BartioChainID:
+		// Bartio does not properly validate deposits index
+		// We skip checks for backward compatibility
+		return nil
+
+	case sp.cs.DepositEth1ChainID() == spec.BoonetEth1ChainID &&
+		slot < math.U64(spec.BoonetFork2Height):
+		// Boonet inherited the bug from Bartio and it may have added some
+		// validators before we activate the fork. So we skip validation
+		// before fork activation
+		return nil
+
+	default:
+		// Verify that outstanding deposits match those listed by contract
+		var depositIndex uint64
+		depositIndex, err = st.GetEth1DepositIndex()
+		if err != nil {
+			return err
+		}
+		expectedStartIdx := depositIndex + 1
+
+		var localDeposits []DepositT
+		localDeposits, err = sp.ds.GetDepositsByIndex(
+			expectedStartIdx,
+			sp.cs.MaxDepositsPerBlock(),
+		)
+		if err != nil {
+			return err
+		}
+
+		sp.logger.Info(
+			"processOperations",
+			"Expected deposit start index", expectedStartIdx,
+			"Expected deposits length", len(localDeposits),
+		)
+
+		if len(localDeposits) != len(deposits) {
+			return errors.Wrapf(
+				ErrDepositsLengthMismatch,
+				"local: %d, payload: %d", len(localDeposits), len(deposits),
+			)
+		}
+
+		for i, sd := range localDeposits {
+			// Deposit indices should be contiguous.
+			if sd.GetIndex().Unwrap() != expectedStartIdx+uint64(i) {
```

Comment:
order should always be maintained. From spec:

```python
def process_deposit(state: BeaconState, deposit: Deposit) -> None:
    ...

    # Deposits must be processed in order
    state.eth1_deposit_index += 1

    apply_deposit(
        state=state,
        pubkey=deposit.data.pubkey,
        withdrawal_credentials=deposit.data.withdrawal_credentials,
        amount=deposit.data.amount,
        signature=deposit.data.signature,
    )
```

Reply:
These checks may be redundant, since these data come from our EVM state and I think they are guaranteed to be in-order. Keeping this for the time being, we will revisit all once merkle root checks are in